File: wcpelib-bukkit/src/main/java/top/wcpe/wcpelib/bukkit/chat/ChatAcceptParameterManager.java
package top.wcpe.wcpelib.bukkit.chat;

import lombok.Data;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import top.wcpe.wcpelib.bukkit.WcpeLib;

import java.util.HashMap;

/**
 * 聊天接受参数
 *
 * @author: WCPE 1837019522@qq.com
 * @create: 2021-05-02 17:06
 */
public class ChatAcceptParameterManager {

    private static HashMap<String, ChatAcceptParameterTask> playerTask = new HashMap<>();

    static {
        Bukkit.getPluginManager().registerEvents(new Listener() {
            @EventHandler(priority = EventPriority.HIGHEST)
            public void playerChat(AsyncPlayerChatEvent e) {
                ChatAcceptParameterTask chatAcceptParameterTask = playerTask.get(e.getPlayer().getName());
                if (chatAcceptParameterTask == null) return;
                if (chatAcceptParameterTask.getTimeStamp() < System.currentTimeMillis()) {
                    playerTask.remove(e.getPlayer().getName());
                    return;
                }
                e.setCancelled(true);
                if (chatAcceptParameterTask.getCancelString().equals(e.getMessage())) {
                    e.getPlayer().sendMessage(chatAcceptParameterTask.getCancelTipString());
                    playerTask.remove(e.getPlayer().getName());
                    return;
                }
                if (chatAcceptParameterTask.getJudge().judge(e.getMessage())) {
                    chatAcceptParameterTask.getJudgeSuccessTask().run(e.getPlayer(), e.getMessage());
                    playerTask.remove(e.getPlayer().getName());
                    return;
                }
                chatAcceptParameterTask.getJudgeFailTask().run(e.getPlayer(), e.getMessage());
                e.getPlayer().sendMessage(chatAcceptParameterTask.getTipString());

            }
        }, WcpeLib.getInstance());

    }

    public static void putChatAcceptParameterTask(Player p, long time, String tipString, String cancelString, String cancelTipString, ChatAcceptParameterReturnJudge judge, ChatAcceptParameterJudgeSuccessTask judgeSuccessTask, ChatAcceptParameterJudgeFailTask judgeFailTask) {
        p.sendMessage(tipString);
        playerTask.put(p.getName(), new ChatAcceptParameterTask(System.currentTimeMillis() + time * 1000, tipString, cancelString, cancelTipString, judge, judgeSuccessTask, judgeFailTask));
    }

    @Data
    private static class ChatAcceptParameterTask {
        public ChatAcceptParameterTask(long timeStamp, String tipString, String cancelString, String cancelTipString, ChatAcceptParameterReturnJudge judge, ChatAcceptParameterJudgeSuccessTask judgeSuccessTask, ChatAcceptParameterJudgeFailTask judgeFailTask) {
            this.timeStamp = timeStamp;
            this.tipString = tipString;
            this.cancelString = cancelString;
            this.cancelTipString = cancelTipString;
            this.judge = judge;
            this.judgeSuccessTask = judgeSuccessTask;
            this.judgeFailTask = judgeFailTask;
        }

        private long timeStamp;
        private String tipString;
        private String cancelString;
        private String cancelTipString;
        private ChatAcceptParameterReturnJudge judge;
        private ChatAcceptParameterJudgeSuccessTask judgeSuccessTask;
        private ChatAcceptParameterJudgeFailTask judgeFailTask;
    }


    @FunctionalInterface
    public interface ChatAcceptParameterReturnJudge {
        boolean judge(String message);
    }

    @FunctionalInterface
    public interface ChatAcceptParameterJudgeSuccessTask {
        void run(Player p, String message);
    }

    @FunctionalInterface
    public interface ChatAcceptParameterJudgeFailTask {
        void run(Player p, String message);
    }
}

File: wcpelib-bukkit/src/main/kotlin/top/wcpe/wcpelib/bukkit/chat/ChatAcceptParameterManager.kt
package top.wcpe.wcpelib.bukkit.chat

import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.AsyncPlayerChatEvent
import top.wcpe.wcpelib.bukkit.WcpeLib
import top.wcpe.wcpelib.bukkit.chat.entity.ChatAcceptParameterTask

/**
 * 由 WCPE 在 2021/12/28 0:40 创建
 *
 * Created by WCPE on 2021/12/28 0:40
 *
 * Github: https://github.com/wcpe
 *
 * QQ: 1837019522
 *
 * @author WCPE
 */
object ChatAcceptParameterManager {

    @JvmStatic
    val playerTask: MutableMap<String, ChatAcceptParameterTask> = mutableMapOf()


    init {
        Bukkit.getPluginManager().registerEvents(object : Listener {
            @EventHandler(priority = EventPriority.HIGH)
            fun playerChatEvent(e: AsyncPlayerChatEvent) {
                playerTask[e.player.name]?.let {
                    if (it.timeStamp < System.currentTimeMillis()) {
                        playerTask.remove(e.player.name)
                        return
                    }

                    e.isCancelled = true

                    if (it.cancelJudgeTask(e.player, e.message)) {
                        it.cancelSuccessTask(e.player, e.message)
                        playerTask.remove(e.player.name)
                        return
                    }
                    if (it.judge(e.player, e.message)) {
                        it.judgeTrueTask(e.player, e.message)
                        playerTask.remove(e.player.name)
                        return
                    }
                    it.judgeFalseTask(e.player, e.message)
                }
            }
        }, WcpeLib.instance)
    }

    fun putChatAcceptParameterTask(
        player: Player,
        inputTime: Long,
        cancelJudgeTask: (player: Player, message: String) -> Boolean,
        cancelSuccessTask: (player: Player, message: String) -> Unit,
        judge: (player: Player, message: String) -> Boolean,
        judgeTrueTask: (player: Player, message: String) -> Unit,
        judgeFalseTask: (player: Player, message: String) -> Unit
    ): Long {
        playerTask[player.name]?.let {
            val currentTimeMillis = System.currentTimeMillis()
            if (it.timeStamp > currentTimeMillis) {
                return it.timeStamp - currentTimeMillis
            }
        }
        playerTask[player.name] = ChatAcceptParameterTask(
            System.currentTimeMillis() + inputTime * 1000,
            cancelJudgeTask,
            cancelSuccessTask,
            judge,
            judgeTrueTask,
            judgeFalseTask
        )
        return -1
    }

}